File: firebase/functions/test/scoringRules.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  calculateHandicap,
  recordScore,
  summarizeStandalone,
  validateScore
} from "../src/domain/scoringRules.ts";

test("standalone fixture totals and winner", () => {
  const summary = summarizeStandalone([
    {
      participantId: "player-a",
      displayName: "Avery",
      scoresByInning: { 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9 }
    },
    {
      participantId: "player-b",
      displayName: "Blake",
      scoresByInning: { 1: 0, 2: 3, 3: 3, 4: 4, 5: 4, 6: 5, 7: 6, 8: 7, 9: 8 }
    }
  ]);

  assert.equal(summary.participantSummaries.find((item) => item.participantId === "player-a")?.total, 45);
  assert.equal(summary.participantSummaries.find((item) => item.participantId === "player-b")?.total, 40);
  assert.deepEqual(summary.winnerIds, ["player-a"]);
  assert.equal(summary.margin, 5);
});

test("invalid scores are rejected", () => {
  assert.throws(() => validateScore(-1));
  assert.throws(() => validateScore(10));
  assert.throws(() => recordScore({ participantId: "a", displayName: "A", scoresByInning: {} }, 1, 10));
});

test("handicap rounding matches fixture", () => {
  const averages = { higher: 72, lower: 63.4 };
  assert.equal(calculateHandicap(averages, 80, "nearest").handicap, 7);
  assert.equal(calculateHandicap(averages, 80, "floor").handicap, 6);
  assert.equal(calculateHandicap(averages, 80, "ceiling").handicap, 7);
  assert.ok(Math.abs(calculateHandicap(averages, 80, "decimal").handicap - 6.88) < 0.001);
});

File: android/app/src/test/java/com/rbdarts/core/domain/ScoringRulesTest.kt
package com.rbdarts.core.domain

import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.Assert.assertThrows

class ScoringRulesTest {
    @Test
    fun standaloneBasicTotals() {
        val participants = listOf(
            ParticipantScorecard("player-a", "Avery", mapOf(1 to 1, 2 to 2, 3 to 3, 4 to 4, 5 to 5, 6 to 6, 7 to 7, 8 to 8, 9 to 9)),
            ParticipantScorecard("player-b", "Blake", mapOf(1 to 0, 2 to 3, 3 to 3, 4 to 4, 5 to 4, 6 to 5, 7 to 6, 8 to 7, 9 to 8))
        )

        val summary = ScoringRules.summarizeStandalone(participants)

        assertThat(summary.participantSummaries.first { it.participantId == "player-a" }.total).isEqualTo(45)
        assertThat(summary.participantSummaries.first { it.participantId == "player-b" }.total).isEqualTo(40)
        assertThat(summary.winnerIds).containsExactly("player-a")
        assertThat(summary.margin).isEqualTo(5)
    }

    @Test
    fun invalidScoresAreRejected() {
        assertThrows(ScoreValidationException::class.java) { ScoringRules.validateScore(-1) }
        assertThrows(ScoreValidationException::class.java) { ScoringRules.validateScore(10) }
    }

    @Test
    fun handicapRoundingMatchesFixture() {
        val averages = mapOf("higher" to 72.0, "lower" to 63.4)

        assertThat(ScoringRules.handicap(averages, 80.0, HandicapRoundingRule.Nearest).handicap).isEqualTo(7.0)
        assertThat(ScoringRules.handicap(averages, 80.0, HandicapRoundingRule.Floor).handicap).isEqualTo(6.0)
        assertThat(ScoringRules.handicap(averages, 80.0, HandicapRoundingRule.Ceiling).handicap).isEqualTo(7.0)
        assertThat(ScoringRules.handicap(averages, 80.0, HandicapRoundingRule.Decimal).handicap).isWithin(0.001).of(6.88)
    }
}
